### ISSUE TITLE
Allow use of native system file picker dialog

### ISSUE THREAD

**Issue by gingerbeardman** [godotengine/godot-proposals#1123]
**Describe the project you are working on:** 

- Pixelorama https://github.com/Orama-Interactive/Pixelorama

**Describe the problem or limitation you are having in your project:** 

1. godot file picker does not work in sandboxed macOS apps 
2. godot file picker does not offer capabilities and conveniences of the system file picker

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

regarding the above issues, allowing use of system file picker:
- is required for 1
- would alleviate 2

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

similar to these existing libraries, but cross platform and in core:
- https://github.com/bruvzg/mac_file_panels
- https://github.com/GodotExplorer/NativeDialogs

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

Sadly not, see #13177 for history

**Is there a reason why this should be core and not an add-on in the asset library?:**

- it is required for sandboxed macOS apps, so core requirement is the only real solution

**Comment by OverloadedOrama** [comment]
This is also necessary for the HTML5 platform, as you can't access the native file system of the device with the FileDialog when running a game/app on browser.

**Comment by gingerbeardman** [comment]
@OverloadedOrama I've added that to the above proposal, thanks!

**Comment by umarcor** [comment]
> FTR, in [Orama-Interactive/Pixelorama](https://github.com/Orama-Interactive/Pixelorama) they work around it with [Pukkah/HTML5-File-Exchange-for-Godot](https://github.com/Pukkah/HTML5-File-Exchange-for-Godot).

**Comment by creikey** [comment]
For those interested in implementing this it seems like this is a good resource for implementing it on windows: https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifiledialog

**Comment by creikey** [comment]
Although I see no reason why https://github.com/mlabbe/nativefiledialog cannot be pulled into third party

**Comment by Calinou** [comment]
This library could be used to implement native file dialogs *and* notifications: https://github.com/samhocevar/portable-file-dialogs

**Comment by Tadaboody** [comment]
> [portable-file-dialogs](https://github.com/samhocevar/portable-file-dialogs) could be used to implement native file dialogs _and_ [desktop notifications](https://github.com/godotengine/godot-proposals/issues/1338).

This looks great! Just note that it doesn't support web yet:
From the readme:
> Experimental support for Emscripten is on its way
I think using it and helping them with a web implementation is the right way to go

**Comment by 98teg** [comment]
This may be solved by [this PR](https://github.com/godotengine/godot/pull/47056), although I still think native file dialogs would be a great addition.

**Comment by bruvzg** [comment]
> This may be solved by this PR, although I still think native file dialogs would be a great addition.

I have replaced it with the full native dialog implementation -  godotengine/godot#47499

> portable-file-dialogs could be used to implement native file dialogs and desktop notifications.

It won't work with the sandboxed macOS apps, since it is relaying on customs scripts, which the sandboxed app won't be able to execute, and even if it was, it won't share permissions with the main app and preserve it across the sessions.

**Comment by andrea-calligaris** [comment]
I would recommend labeling this "high priority". Other than mine, I can imagine a lot of GUI projects currently stuck because of this.
If you are making a mobile app, in most cases you require the user to load a file, e.g. to restore a backup of user's data; but even in games, allowing the user to load a _"skin.png"_ file or stuff like that is pretty common.
And right now it's simply impossible because `FileDialog` is ultra-bugged, and even more bugged in mobile.

**Comment by Zami77** [comment]
Throwing my 2 cents in that I'd like to see this core as well. Currently running into this issue for a UI I developed that involves uploading a file to be scanned by a backend. Works great on windows export, unfortunately not possible on the HTML5 build.

**Comment by spindlebink** [comment]
I took a stab at this and it looks like `portable-file-dialogs` relies pretty exclusively on the C++ stdlib. I remember reading a while back that Godot stays away from std--is that still the case?

**Comment by spindlebink** [comment]
Awesome. I'll double check for no exceptions when I'm back at dev.

I did have a couple of wrapping-up questions if you've got the time--
* Is there a guide somewhere for editor-specific stuff in a class? I still need to add an icon, editor description, tooltips, etc. and I'd like to get that done while still developing the initial version, since it's often less likely to be completed later on.
* Do you have a suggestion for a way to add an emitted signal to this class? The library only provides checking for whether the window has closed, so my first instinct would be just to add a check during `_process` and emit a `window_dismissed` (or similar) signal when it detects the change. Still, that feels a little inelegant, and if I'm missing a better way (maybe a C++-side `await` equivalent or a structure for it?) that'd be good to get info on.

**Comment by Calinou** [comment]
< Is there a guide somewhere for editor-specific stuff in a class? I still need to add an icon, editor description, tooltips, etc. and I'd like to get that done while still developing the initial version, since it's often less likely to be completed later on.

See [Editor icons](https://docs.godotengine.org/en/stable/development/editor/creating_icons.html) and [editor style guide](https://docs.godotengine.org/en/stable/development/editor/editor_style_guide.html) in the documentation.

Adding tooltips to properties is done by exposing them to the scripting API, then [updating the class reference](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html).

**Comment by octetdev2** [comment]
Tangental but native color pickers are useful too as they lets you color pick anything on screen, rather than just the things Godot draws.

**Comment by 98teg** [comment]
The plugin I mentioned earlier is now finished and available at the [asset library](https://godotengine.org/asset-library/asset/1344). I hope you find it useful while this functionality gets added to Godot.

**Comment by 9gay** [comment]
if you use xdg-desktop-portal you're gonna break [netbsd](https://pkgsrc.se/search.php?so=portal) and [openbsd](https://openports.se/search.php?so=portal) support (well, it's possible for them to port portal, but without flatpak it doesn't make sense)

when it comes to linux and bsd it's complicated to tell what is native, if you use kde, lxqt, deepin or lumina then you surely get qt file picker, in all other cases you unfortunately (if you use raw window manager but you like qt more than gtk) get gtk file picker

considering people use raw window managers to save resources, tcl/tk is a good option, because it's extremely easy to implement support for it and it uses native file picker on windows and macos, but implements its own on linux and bsd, you don't even need to use it as a C library (although it's designed to be embedded that way) because you can just use it from command line (assuming it is installed in PATH) `echo 'wm withdraw .; puts [tk_getOpenFile]; exit' | wish` (other methods are `tk_getSaveFile` and `tk_chooseDirectory`) and it returns either empty string if user pressed cancel or path to file/directory if user pressed open/save/ok

**Comment by ionutrogojan** [comment]
Is this still being considered? I'm thinking of building an application with Godot and I would really like the native file dialog.

**Comment by Calinou** [comment]
> Is this still being considered? I'm thinking of building an application with Godot and I would really like the native file dialog.

There's a pull request implementing this feature for macOS: https://github.com/godotengine/godot/pull/47499

This hasn't been implemented for other platforms yet.

**Comment by Anutrix** [comment]
Afaik, native file selection dialog support is implemented for Windows, Linux and macOS.
What's status on Android, Web and iOS?

**Comment by bruvzg** [comment]
> Afaik, native file selection dialog support is implemented for Windows, Linux and macOS.

It is implemented on Windows, Linux and macOS, but only for projects and for `file system` access, Godot Editor currently does no use them. Some of the editor file dialogs have custom controls added. Native dialogs can display extra options on all supported platforms, and it should be sufficient for the editor needs (I'm working on adding support for it).

> What's status on Android, Web and iOS?

Not implemented on iOS, Android, native file selectors can be implemented, and probably will be much more convenient for users due to limited file system access. Web does not have any access to the real file system.

**Comment by OverloadedOrama** [comment]
>  Web does not have any access to the real file system.

It's already possible to call custom JavaScript code that brings up native file dialogs that have access to the device's file system, see https://github.com/Pukkah/HTML5-File-Exchange-for-Godot. Would something like this not be possible to be implemented in Godot?

**Comment by solarite1797** [comment]
bump

**Comment by syntaxerror247** [comment]
Hi everyone, I've created a draft for the [Android system file picker](https://github.com/godotengine/godot/pull/98350). It would be great if you could test it and let me know if you encounter any issues or errors.

**Comment by vgezer** [comment]
> Native file dialogs are now supported for all desktop platforms via FileDialog's `use_native_dialog` property.

Unfortunately, it looks like it still uses Godot on Web, even `force native` and `use native` are enabled.

[nonnativewindows.zip](https://github.com/user-attachments/files/17940205/nonnativewindows.zip)

![Image](https://github.com/user-attachments/assets/8d6df6c7-a94f-4fdd-9f6c-7b26af5c2150)


Created with Godot 4.4.dev4 / and dev5 on Windows.

Tried with Chrome and Firefox...

**Comment by KoBeWi** [comment]
Web is not a desktop platform.

**Comment by vgezer** [comment]
> Web is not a desktop platform.

Ops, sorry. That also explains why there exist plugins :)

**Comment by jhaleit** [comment]
Yes pls support iOS soon.. I can’t complete my project without this feature..

**Comment by Calinou** [comment]
@jhaleit Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by jhaleit** [comment]
Just wanted to give an update.. I did a thing here for iOS: https://github.com/godotengine/godot/pull/109484
